src/core/dfn.js

```diff
@@ -36,6 +102,21 @@ export function run() {
   }
   sub("plugins-done", addContractDefaults);
 }
+
+function validateDefinition(dfn, linkingText, type) {
+  const entry = knownTypesMap.get(type);
+  if (entry.requiresFor && !dfn.dataset.dfnFor) {
+    const msg = docLink`Definition of type "\`${type}\`" requires a ${"[data-dfn-for]"} attribute.`;
+    const { associateWith } = entry;
+    const hint = docLink`Use a ${"[data-dfn-for]"} attribute to associate this with ${associateWith}.`;
+    showError(msg, name, { hint, elements: [dfn] });
+  }
+
+  if (entry.validator) {
+    entry.validator(linkingText, type, dfn, name);
```

Comment:
I wonder if it would make sense to throw from validators and showError here?

Reply:
The errors might be quire custom per validator, so I'd like to just leave them inside the validators themselves.